cpp/comms/include/cuML_comms.hpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <nccl.h>
+
+#ifdef WITH_UCX
+#include <ucp/api/ucp.h>
+#endif
+
+#include <cuML.hpp>
+
+namespace ML {
+
+/**
+ * @brief Given initialized comms handles for NCCL and UCP, this function builds a
+ * cumlCommunicator object and injects it into the given cumlHandle instance.
+ * @param handle the cuml handle to inject a new communicator instance into
+ * @param comm initialized nccl communicator
+ * @param ucp_worker the ucp_worker for the current initialized ucp context
+ * @param eps an array of endpoints to the other ucp workers in the cluster
+ * @param size the size of the cluster (number of elements in eps)
+ * @param rank rank of the current worker
+ */
+#ifdef WITH_UCX
+void inject_comms(cumlHandle &handle, ncclComm_t comm, ucp_worker_h ucp_worker,
+                  ucp_ep_h *eps, int size, int rank);
+#endif
+
+/**
+ * @brief Given an initialized comms handle for NCCL, this function builds a
+ * cumlCommunicator object and injects it into the given cumlHandle instance.
+ * The underlying cumlCommunicator will only have support for collective
+ * communications functions.
+ * @param handle the cuml handle to inject a new communicator instance into
+ * @param comm initialized nccl communicator
+ * @param size the size of the cluster
+ * @param rank rank of the current worker
+ */
+void inject_comms(cumlHandle &handle, ncclComm_t comm, int size, int rank);
+
+/**
```

Comment:
this also seems to be declared in cuML_comms_py.hpp?

Reply:
In order to not have the Python depend directly on UCX structs, and having to re-define these structs in cython is nontrivial, I have created `*_py` variants for the public functions that Cython uses. I needed to put them in a separate header that does not include any ucp dependencies. This technique is done inside the ucx-py project as well.

Reply:
Ah, i see the duplicate declaration. You are right, I missed pulling these out when I created the new header.

Reply:
Could you only use opaque, e.g. `void*` in the definitions of `inject_comms` if so the header would not have any UCX or other dependencies and you could avoid this duplication. As those function could be used both from C++ and Python/Cython.

Reply:
I don't mind moving the header file for the Python-specific functions out of the `include/` directory, but I'm really wanting to keep the Python vs C++ concerns separate here, so as to keep the C++ (non-python) use of the API as clean as possible and not introduce unnecessary complexity in the public-facing functions.

I'm finding the use of these wrapper functions to be quite common in libraries which use Cython. This is why I've separated the python-specific functions into their own header file.

Reply:
Agree. I think my comment here was resolved by your explanation why you can't simply cast pointers to an underlying array.